### ISSUE TITLE
[FEATURE] batch inference / GPU parallelism

### ISSUE THREAD

**Issue by ideasbyjin** [brennanaba/ImmuneBuilder#8]
Hi Brennan (and other OPIGlets working on this),

Great tool! Couple of topics for you :) 

1. Batch infererence
I just noted that in the publication you're using a bsz of 64 for training, and I'm certain you can do batch inference on this. The code as-is doesn't provide any batch inference capabilities and would be good to have this for other users. Much of the "USP" of ABodyBuilder2 is on how you can model structures at large scale, but this current implementation only allows single sequence inference. It's not entirely clear how you handle e.g. differences in lengths – how are pads taken into account, or do you always make sure your batch has identical length sequences in a batch? How do pads affect node and edge embeddings, etc. 

2. GPU/data parallelism - more a question than suggestion
I noted that you're putting -all- the models into one GPU. Granted, they're quite small, but doesn't it make more sense to distribute to n GPUs instead? This could reduce memory footprint per GPU, and coupled with suggestion 1., you could really start to batch infer lots of sequences at once.

**Comment by ideasbyjin** [comment]
Oof! Doesn't that mean you needed like 64 steps of gradient accumulation? I feel for you man

As for multi-GPU, so this is where it gets interesting. I essentially wanted to see if we can parallelise the inference step rather than doing it serially, then taking the average. In my head you can do this with `nn.DataParallel` but that assumes your model is essentially copied across the GPUs as opposed to having four different models (as in your case)... needs a think.

**Comment by edikedik** [comment]
For those who still need it, it's possible to use "batching" via GPU parallelism with `ray` as shown below. You can easily adapt it to your needs and hardware. For instance, I can process around 10 structures in parallel on a single RTX 4090 card.

```python
from collections import abc
from itertools import cycle
from pathlib import Path

import numpy as np
import ray
import torch
import lXtractor.chain as lxc  # do not remove
from ImmuneBuilder.ABodyBuilder2 import Antibody
from ImmuneBuilder.models import StructureModule
from ImmuneBuilder.sequence_checks import number_single_sequence
from ImmuneBuilder.util import are_weights_ready, download_file, get_encoding
from lXtractor.util import apply

EMBED_DIM = {
    "antibody_model_1": 128,
    "antibody_model_2": 256,
    "antibody_model_3": 256,
    "antibody_model_4": 256,
}

MODEL_URLS = {
    "antibody_model_1": "https://zenodo.org/record/7258553/files/antibody_model_1?download=1",
    "antibody_model_2": "https://zenodo.org/record/7258553/files/antibody_model_2?download=1",
    "antibody_model_3": "https://zenodo.org/record/7258553/files/antibody_model_3?download=1",
    "antibody_model_4": "https://zenodo.org/record/7258553/files/antibody_model_4?download=1",
}


def join_numbered_seq(seq) -> str:
    return "".join([x[1] for x in seq])


def prepare_ab(
    ab_seq: tuple[str, str]
) -> tuple[str, str, dict[str, list], dict[str, str], np.ndarray]:
    """
    Prepare a single input antibody sequence.
    """
    vh = number_single_sequence(ab_seq[0], "H")
    vl = number_single_sequence(ab_seq[1], "L")
    d = {"H": vh, "L": vl}
    dj = {"H": join_numbered_seq(vh), "L": join_numbered_seq(vl)}
    enc = get_encoding(dj)
    return vh, vl, d, dj, enc


def prepare_inputs(
    ab_seqs: abc.Iterable[tuple[str, str]],
    num_proc: int,
    verbose: bool = True,
    desc: str = "Preparing inputs",
) -> abc.Iterator[tuple[str, str, dict[str, list], dict[str, str], np.ndarray]]:
    """
    Prepare inputs for ABodyBuilder2.
    """
    yield from apply(
        prepare_ab, ab_seqs, verbose, desc, num_proc=num_proc, use_joblib=True
    )


class ModelWorker:
    """
    An actor to be wrapped in `ray.remote`.
    """
    def __init__(self, model_refs):
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.models = [ray.get(model_ref) for model_ref in model_refs]
        for model in self.models:
            model.to(self.device)
            model.eval()

    def predict(self, encoding, full_seq):
        encoding = encoding.to(self.device)
        with torch.no_grad():
            outputs = [model(encoding, full_seq) for model in self.models]
            return [
                (atoms.detach().cpu(), features.detach().cpu())
                for atoms, features in outputs
            ]


class ABodyBuilder2ray:
    def __init__(
        self,
        model_ids=(1, 2, 3, 4),
        weights_dir=None,
        numbering_scheme="imgt",
        gpu_id: int | None = None,
    ):
        self.device = "cuda" if torch.cuda.is_available() else "cpu"
        if self.device == "cuda" and gpu_id is not None:
            self.device += f":{gpu_id}"
        self.scheme = numbering_scheme

        if weights_dir is None:
            weights_dir = Path(__file__).resolve().parent.parent / "abb2"
            weights_dir.mkdir(exist_ok=True, parents=True)

        # Initialize models and create shared references
        self.model_refs = []
        for id in model_ids:
            model_file = f"antibody_model_{id}"
            model = StructureModule(rel_pos_dim=64, embed_dim=EMBED_DIM[model_file]).to(
                self.device
            )
            weights_path = weights_dir / model_file

            if not are_weights_ready(weights_path):
                print(f"Downloading weights for {model_file}...", flush=True)
                download_file(MODEL_URLS[model_file], weights_path)

            model.load_state_dict(torch.load(weights_path, map_location=self.device))
            model.to(torch.get_default_dtype())
            model.eval()

            # Create a shared reference to the model
            self.model_refs.append(ray.put(model))

    def create_workers(
        self, num_workers: int, cpu_per_worker: int, gpu_frac_per_worker: float
    ):
        model_remote = ray.remote(ModelWorker).options(
            num_cpus=cpu_per_worker, num_gpus=gpu_frac_per_worker
        )
        return [model_remote.remote(self.model_refs) for _ in range(num_workers)]

    def predict(
        self,
        sequences: abc.Iterable[tuple[str, str]],
        num_proc: int,
        num_workers: int,
        cpu_per_worker: int,
        gpu_frac_per_worker: float,
    ) -> abc.Iterator[Antibody]:
        workers = self.create_workers(num_workers, cpu_per_worker, gpu_frac_per_worker)
        worker_cycle = cycle(workers)

        futures = []

        inputs = list(prepare_inputs(sequences, num_proc))

        for input_datum in inputs:
            anarci_numbering = input_datum[2]
            encoding = torch.tensor(input_datum[-1], dtype=torch.get_default_dtype())
            full_seq = input_datum[-2]["H"] + input_datum[-2]["L"]

            worker = next(worker_cycle)
            future = worker.predict.remote(encoding, full_seq)
            futures.append((future, anarci_numbering))

        for future, anarci_numbering in futures:
            output = ray.get(future)
            yield Antibody(anarci_numbering, output)
```
